### PR TITLE
Query using Labels

### PR DESCRIPTION
**Pull request by ruchim**
- [x] Update Swagger
- [x] Centaur tests
- [x] Update readme
- [x] Update changelog

### CODE REVIEW

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala [hunk 1]
```diff
@@ -113,18 +132,21 @@ trait MetadataSlickDatabase extends MetadataSqlDatabase {
   }
 
   override def refreshMetadataSummaryEntries(metadataKey1: String, metadataKey2: String, metadataKey3: String,
-                                             metadataKey4: String, buildUpdatedSummary:
+                                             metadataKey4: String, metadataLabelKey: String, buildUpdatedSummary:
```
Comment (cjllanwarne) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: ToL: Can these other keys possibly be given better names?

File: database/sql/src/main/scala/cromwell/database/slick/tables/CustomLabelEntryComponent.scala [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    } yield (customLabelEntry.customLabelKey, customLabelEntry.customLabelValue)
+  )
+
+  def existsWorkflowIdLabelKeyAndValue(workflowId: Rep[String])
+                                      (labelKeylabelValue: (String, String)): Rep[Boolean] = {
+    customLabelEntries.filter(customLabelEntry =>
+      customLabelEntry.workflowExecutionUuid === workflowId &&
+        customLabelEntry.customLabelKey === labelKeylabelValue._1 &&
+        customLabelEntry.customLabelValue === labelKeylabelValue._2
+    ).exists
```
Comment (cjllanwarne) [inline database/sql/src/main/scala/cromwell/database/slick/tables/CustomLabelEntryComponent.scala]: ToL: This one seems to be a totally different style to the previous query, presumably necessarily?

File: services/src/main/scala/cromwell/services/metadata/WorkflowQueryParameters.scala [hunk 3]
```diff
@@ -60,10 +61,13 @@ object WorkflowQueryParameters {
       rawParameters groupBy { case (key, _) => key.toLowerCase.capitalize }
 
     val Seq(startDate, endDate) = Seq(StartDate, EndDate) map { _.validate(valuesByCanonicalCapitalization) }
+
     val Seq(statuses, names, ids) = Seq(Status, Name, WorkflowQueryKey.Id) map {
       _.validate(valuesByCanonicalCapitalization)
     }
 
+    val Seq(labels) = Seq(WorkflowQueryKey.LabelKeyValue) map { _.validate(valuesByCanonicalCapitalization) }
```
Comment (cjllanwarne) [inline services/src/main/scala/cromwell/services/metadata/WorkflowQueryParameters.scala]: Where `canonicalCapitalization` for labels means... no capitalization?

File: services/src/test/scala/cromwell/services/metadata/WorkflowQueryParametersSpec.scala [hunk 4]
```diff
@@ -142,6 +146,38 @@ class WorkflowQueryParametersSpec extends WordSpec with Matchers {
       }
     }
 
+    "reject bad label regex" in {
+      val badLabelRegex = "0-label-key"
```
Comment (cjllanwarne) [inline services/src/test/scala/cromwell/services/metadata/WorkflowQueryParametersSpec.scala]: I don't think `regex` is needed in this test name, and this could just be `badLabelKey`

File: src/bin/travis/testCentaurJes.sh [hunk 5]
```diff
@@ -26,19 +26,36 @@ printTravisHeartbeat() {
...
 killTravisHeartbeat() {
     if [ -n "${TRAVIS_HEARTBEAT_PID+set}" ]; then
         kill ${TRAVIS_HEARTBEAT_PID} || true
     fi
 }
 
 exitScript() {
+    echo "CROMWELL LOG"
+    cat log.txt
     killTravisHeartbeat
```
Comment (cjllanwarne) [inline src/bin/travis/testCentaurJes.sh]: I think you want to kill the cromwell log tail here too

**Comment by cjllanwarne** [comment]
Just to double check, the custom and default labels all end up going to the summary tables, right? I can't remember that I saw that

File: database/migration/src/main/resources/changesets/custom_label_entry.xml [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+                                 baseColumnNames="WORKFLOW_EXECUTION_UUID"
+                                 referencedTableName="WORKFLOW_METADATA_SUMMARY_ENTRY"
+                                 referencedColumnNames="WORKFLOW_EXECUTION_UUID"
+                                 onDelete="CASCADE"
+        />
+    </changeSet>
+
+    <changeSet author="rmunshi" id="custom_label_entry_index">
+        <addUniqueConstraint constraintName="UC_CUSTOM_LABEL_ENTRY_CLK_CLV_WEU"
+                             tableName="CUSTOM_LABEL_ENTRY" columnNames="CUSTOM_LABEL_KEY, CUSTOM_LABEL_VALUE, WORKFLOW_EXECUTION_UUID" />
```
Comment (mcovarr) [inline database/migration/src/main/resources/changesets/custom_label_entry.xml]: Yup that sounds right

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala [hunk 7]
```diff
@@ -93,6 +94,24 @@ trait MetadataSlickDatabase extends MetadataSqlDatabase {
     } yield ()
   }
 
+  private def updateCustomLabelEntry(metadataEntry: MetadataEntry)
```
Comment (mcovarr) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: Summary tables are CRUD not CRAP

Comment (mcovarr) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: Perhaps the comment and/or naming could be more clear that this is a summary table.

Comment (cjllanwarne) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: Sorry @mcovarr, I didn't mean to offend your tables! :)

File: database/migration/src/main/resources/changesets/custom_label_entry.xml [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+    <changeSet id="custom_LABEL_entry" author="rmunshi">
+        <createTable tableName="CUSTOM_LABEL_ENTRY">
+            <column autoIncrement="true" name="CUSTOM_LABEL_ENTRY_ID" type="BIGINT">
+                <constraints primaryKey="true" primaryKeyName="PK_CUSTOM_LABEL_ENTRY"/>
+            </column>
+            <column name="CUSTOM_LABEL_KEY" type="VARCHAR(255)">
+                <constraints nullable="false"/>
+            </column>
+            <column name="CUSTOM_LABEL_VALUE" type="VARCHAR(255)">
```
Comment (geoffjentry) [inline database/migration/src/main/resources/changesets/custom_label_entry.xml]: Aside: Given that labels are rapidly evolving beyond "crap that we push through to JES" that statement might not hold going forward

**Comment by geoffjentry** [comment]
Could you add some doc to the code to explain what some of the higher level concepts are and what's going on?

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala (hunk 1 above)
Comment (mcovarr) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: They used to have better names and were changed to this for reasons I don't remember (not by me)

File: database/sql/src/main/scala/cromwell/database/slick/tables/CustomLabelEntryComponent.scala (hunk 2 above)
Comment (mcovarr) [inline database/sql/src/main/scala/cromwell/database/slick/tables/CustomLabelEntryComponent.scala]: not 100% sure it couldn't be compiled like the others

File: database/migration/src/main/resources/changesets/custom_label_entry.xml [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+                <constraints nullable="false"/>
```
Comment (mcovarr) [inline database/migration/src/main/resources/changesets/custom_label_entry.xml]: values are not nullable?

File: database/migration/src/main/resources/changesets/custom_label_entry.xml [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+        </createTable>
+    </changeSet>
+
+    <changeSet author="rmunshi" id="WMSE_workflow_execution_uuid_foreign_key">
+        <addForeignKeyConstraint constraintName="FK_CUSTOM_LABEL_ENTRY_WORKFLOW_EXECUTION_UUID"
+                                 baseTableName="CUSTOM_LABEL_ENTRY"
+                                 baseColumnNames="WORKFLOW_EXECUTION_UUID"
+                                 referencedTableName="WORKFLOW_METADATA_SUMMARY_ENTRY"
+                                 referencedColumnNames="WORKFLOW_EXECUTION_UUID"
+                                 onDelete="CASCADE"
```
Comment (mcovarr) [inline database/migration/src/main/resources/changesets/custom_label_entry.xml]: Why is the `onDelete` here?  AFAIK nothing should ever be deleted from the referenced summary table.

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala [hunk 11]
```diff
@@ -93,6 +94,24 @@ trait MetadataSlickDatabase extends MetadataSqlDatabase {
...
+                                    (implicit ec: ExecutionContext): DBIO[Unit] = {
+    val labelKey = metadataEntry.metadataKey.split("\\:", 2)(1)
```
Comment (mcovarr) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: maybe a comment here as to the pattern you're matching

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala [hunk 12]
```diff
@@ -113,18 +132,21 @@ trait MetadataSlickDatabase extends MetadataSqlDatabase {
...
     val action = for {
       previousMetadataEntryIdOption <- getSummaryStatusEntryMaximumId(
         "WORKFLOW_METADATA_SUMMARY_ENTRY", "METADATA_ENTRY")
       previousMetadataEntryId = previousMetadataEntryIdOption.getOrElse(0L)
       metadataEntries <- dataAccess.metadataEntriesForIdGreaterThanOrEqual((
-        previousMetadataEntryId + 1L, metadataKey1, metadataKey2, metadataKey3, metadataKey4)).result
-      metadataByWorkflowUuid = metadataEntries.groupBy(_.workflowExecutionUuid)
+        previousMetadataEntryId + 1L, metadataKey1, metadataKey2, metadataKey3, metadataKey4, likeMetadataLabelKey)).result
+      metadataByWorkflowUuid = metadataEntries.filterNot(_.metadataKey.contains(metadataLabelKey)).groupBy(_.workflowExecutionUuid)
+      customLabelEntries = metadataEntries.filter(_.metadataKey.contains(metadataLabelKey))
```
Comment (mcovarr) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: The naming of `metadataByWorkflowUuid` doesn't make it clear that this collection excludes labels.

You might build both collections with a single `partition`:

`(customLabelEntries, metadataWithoutLabels) = metadataEntries.partition(_.metadataKey.contains(metadataLabelKey))
`

Though you'd still need to run the `groupBy` on the non-label metadata.

File: database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowMetadataSummaryEntryComponent.scala [hunk 13]
```diff
@@ -82,10 +84,13 @@ trait WorkflowMetadataSummaryEntryComponent {
       val workflowStatusFilter = workflowStatuses.
         map(workflowStatus => workflowMetadataSummaryEntry.workflowStatus.fold(ifEmpty = exclude)(_ === workflowStatus)).
         reduceLeftOption(_ || _)
-
+      val labelsFilter = labelKeyLabelValues.
```
Comment (mcovarr) [inline database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowMetadataSummaryEntryComponent.scala]: A comment here would be great

File: services/src/main/scala/cromwell/services/metadata/WorkflowQueryKey.scala [hunk 14]
```diff
@@ -37,16 +40,32 @@ object WorkflowQueryKey {
     override val name = "Name"
 
     override def validate(grouped: Map[String, Seq[(String, String)]]): ErrorOr[List[String]] = {
-
       val values = valuesFromMap(grouped).toList
-      val nels = values map {
+      val nels:List[data.ValidatedNel[String,String]] = values map {
```
Comment (mcovarr) [inline services/src/main/scala/cromwell/services/metadata/WorkflowQueryKey.scala]: Is this ascription now necessary?  If so, space after the colon.

File: services/src/main/scala/cromwell/services/metadata/WorkflowQueryKey.scala [hunk 15]
```diff
@@ -37,16 +40,32 @@ object WorkflowQueryKey {
...
+      val values = valuesFromMap(grouped).toList
+
+      def validateLabelRegex(labelKeyValue: String): ErrorOr[Label] = {
+        labelKeyValue.split("\\:", 2) match {
+          case Array(k, v) => Label.validateLabel(k, v)
+          case other => s"$labelKeyValue".invalidNel
+        }
+      }
+      val nels: List[ErrorOr[Label]] = values map validateLabelRegex
+      sequenceListOfValidatedNels(s"Label values do not match allowed pattern label-key:label-value", nels)
```
Comment (mcovarr) [inline services/src/main/scala/cromwell/services/metadata/WorkflowQueryKey.scala]: Unnecessary `s` on string

File: services/src/main/scala/cromwell/services/metadata/WorkflowQueryKey.scala [hunk 16]
```diff
@@ -106,6 +125,16 @@ sealed trait SeqStringWorkflowQueryKey extends WorkflowQueryKey[Seq[String]] {
   }
 }
 
+sealed trait SeqLabelWorkflowQueryKey extends WorkflowQueryKey[Seq[Label]] {
```
Comment (mcovarr) [inline services/src/main/scala/cromwell/services/metadata/WorkflowQueryKey.scala]: This code looks strikingly similar to that 10 lines above it.  😄   Seems like this could be DRYed up.

File: src/bin/travis/testCentaurJes.sh (hunk 5 above)
Comment (mcovarr) [inline src/bin/travis/testCentaurJes.sh]: Are these changes relevant to this PR?  It looks like useful stuff, I'm just not sure why it's her.

File: engine/src/main/scala/cromwell/webservice/WorkflowJsonSupport.scala [hunk 17]
```diff
@@ -43,7 +43,7 @@ object WorkflowJsonSupport extends DefaultJsonProtocol {
     }
   }
 
-  implicit val workflowQueryResult = jsonFormat5(WorkflowQueryResult)
+  implicit val workflowQueryResult = jsonFormat6(WorkflowQueryResult)
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/webservice/WorkflowJsonSupport.scala]: I know you have checkboxes for updating the README and CHANGELOG, but we may want to make sure our big customers are aware of the change in this format.

File: services/src/test/scala/cromwell/services/metadata/impl/MetadataDatabaseAccessSpec.scala [hunk 18]
```diff
@@ -113,9 +118,13 @@ class MetadataDatabaseAccessSpec extends FlatSpec with Matchers with ScalaFuture
...
       val randomIds = Seq.fill(10)(WorkflowId.randomId().toString)
 
+      val testLabel1 = Label.safeLabel("testing-key-1", "testing-value-1")
+      val testLabel2 = Label.safeLabel("testing-key-2", "testing-value-2")
+      val testLabel3 = Label.safeLabel("testing-key-3", "testing-value-3")
+
       def succeededWorkflowMetadata(id: WorkflowId): Future[Unit] = {
         val workflowKey = MetadataKey(id, jobKey = None, key = null)
-        val keyAndValue = Array(
+        val keyAndValue= Array(
```
Comment (mcovarr) [inline services/src/test/scala/cromwell/services/metadata/impl/MetadataDatabaseAccessSpec.scala]: revert

File: engine/src/main/resources/swagger/cromwell.yaml [hunk 19]
```diff
@@ -264,6 +264,19 @@ paths:
...
+        - name: label
+          required: false
+          in: query
+          type: array
+          items:
+            type: string
+          collectionFormat: multi
+          pattern: ^([a-z][-a-z0-9]*[a-z0-9])?[:]([a-z][-a-z0-9]*[a-z0-9])?$
+          description: >
+            Returns workflows with the specified label keys.  If specified multiple times,
```
Comment (mcovarr) [inline engine/src/main/resources/swagger/cromwell.yaml]: Is it the case that this supports key search and key:value search but not both types in the same query?

**Comment by mcovarr** [comment]
:+1: with some nits picked and assuming all your checkboxes get checked.  😄

File: .travis.yml [hunk 20]
```diff
@@ -19,7 +19,7 @@ before_install:
   - openssl aes-256-cbc -K "$encrypted_5ebd3ff04788_key" -iv "$encrypted_5ebd3ff04788_iv" -in src/bin/travis/resources/jesConf.tar.enc -out jesConf.tar -d || true
 env:
   global:
-    - CENTAUR_BRANCH=develop
+    - CENTAUR_BRANCH=dd376b1831475ae21272d41d8c335e9f8bd9f259
```
Comment (ruchim) [inline .travis.yml]: :+1:

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala [hunk 21]
```diff
@@ -93,6 +94,24 @@ trait MetadataSlickDatabase extends MetadataSqlDatabase {
     } yield ()
   }
 
+  private def updateCustomLabelEntry(metadataEntry: MetadataEntry)
+                                    (implicit ec: ExecutionContext): DBIO[Unit] = {
+    val labelKey = metadataEntry.metadataKey.split("\\:", 2)(1)
+    val labelValue = metadataEntry.metadataValue.toRawString
+    val customLabelEntry = buildCustomLabelEntry(labelKey, labelValue, metadataEntry.workflowExecutionUuid)
+    for {
+      _ <- dataAccess.customLabelEntryIdsAutoInc.insertOrUpdate(customLabelEntry)
```
Comment (danbills) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: if this method returns () then you can get rid of the for comprehension

File: database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowMetadataSummaryEntryComponent.scala [hunk 22]
```diff
@@ -82,10 +84,13 @@ trait WorkflowMetadataSummaryEntryComponent {
...
+        map(labelKeyLabelValue =>
+          existsWorkflowIdLabelKeyAndValue(workflowMetadataSummaryEntry.workflowExecutionUuid)(labelKeyLabelValue)).
```
Comment (danbills) [inline database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowMetadataSummaryEntryComponent.scala]: think you can omit parameter name and use `existsWorkflowIdLabelKeyAndValue(workflowMetadataSummaryEntry.workflowExecutionUuid) _ `

File: database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowMetadataSummaryEntryComponent.scala [hunk 23]
```diff
@@ -55,7 +55,9 @@ trait WorkflowMetadataSummaryEntryComponent {
   )
 
   def filterWorkflowMetadataSummaryEntries(workflowStatuses: Set[String], workflowNames: Set[String],
-                                           workflowExecutionUuids: Set[String], startTimestampOption: Option[Timestamp],
+                                           workflowExecutionUuids: Set[String],
+                                           labelKeyLabelValues: Set[(String,String)],
+                                           startTimestampOption: Option[Timestamp],
```
Comment (danbills) [inline database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowMetadataSummaryEntryComponent.scala]: ToL: this seems like something you could declare separately and combine separately, to keep the pieces small and digestible.  I'm working to have something concrete to suggest but this is what my brain is saying.

File: database/migration/src/main/resources/changesets/custom_label_entry.xml (hunk 9 above)
Comment (cjllanwarne) [inline database/migration/src/main/resources/changesets/custom_label_entry.xml]: Not really... are you thinking labels could just be like tags?

File: database/migration/src/main/resources/changesets/custom_label_entry.xml (hunk 10 above)
Comment (cjllanwarne) [inline database/migration/src/main/resources/changesets/custom_label_entry.xml]: CRU~~D~~

File: services/src/main/scala/cromwell/services/metadata/WorkflowQueryKey.scala (hunk 14 above)
Comment (cjllanwarne) [inline services/src/main/scala/cromwell/services/metadata/WorkflowQueryKey.scala]: Intellij now bothers us to annotate public methods and variables. So, unless we can make this private...?

File: engine/src/main/scala/cromwell/webservice/WorkflowJsonSupport.scala (hunk 17 above)
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/webservice/WorkflowJsonSupport.scala]: Wait, I didn't notice this... why is the query result getting wider? We shouldn't need to be including labels in the query result?

File: services/src/main/scala/cromwell/services/metadata/WorkflowQueryKey.scala (hunk 14 above)
Comment (geoffjentry) [inline services/src/main/scala/cromwell/services/metadata/WorkflowQueryKey.scala]: One can always stop jumping off bridges just because Intellij says to do so ....

Comment (mcovarr) [inline services/src/main/scala/cromwell/services/metadata/WorkflowQueryKey.scala]: I thought Khalid had worked some magic to turn that off.

File: database/migration/src/main/resources/changesets/custom_label_entry.xml (hunk 8 above)
Comment (cjllanwarne) [inline database/migration/src/main/resources/changesets/custom_label_entry.xml]: IMO "what JES will accept" is as good a definition for our labels as anything else, at least until we get conflicting requirements (at which point we could work out how to change it)

File: services/src/main/scala/cromwell/services/metadata/WorkflowQueryKey.scala (hunk 14 above)
Comment (cjllanwarne) [inline services/src/main/scala/cromwell/services/metadata/WorkflowQueryKey.scala]: I don't consider type ascription to be jumping off bridges. Personally I like it

Comment (geoffjentry) [inline services/src/main/scala/cromwell/services/metadata/WorkflowQueryKey.scala]: I was more talking about invoking IntelliJ as a reason to do something

File: database/migration/src/main/resources/changesets/custom_label_entry.xml (hunk 8 above)
Comment (ruchim) [inline database/migration/src/main/resources/changesets/custom_label_entry.xml]: Changed to varchar(63)

File: database/migration/src/main/resources/changesets/custom_label_entry.xml (hunk 10 above)
Comment (ruchim) [inline database/migration/src/main/resources/changesets/custom_label_entry.xml]: Totally right, never delete from summary. Line 29 removed.

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala (hunk 7 above)
Comment (ruchim) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: This is certainly insert only, renamed and updated to reflect reality.

File: database/migration/src/main/resources/changesets/custom_label_entry.xml [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+            <column autoIncrement="true" name="CUSTOM_LABEL_ENTRY_ID" type="BIGINT">
+                <constraints primaryKey="true" primaryKeyName="PK_CUSTOM_LABEL_ENTRY"/>
+            </column>
+            <column name="CUSTOM_LABEL_KEY" type="VARCHAR(255)">
+                <constraints nullable="false"/>
+            </column>
+            <column name="CUSTOM_LABEL_VALUE" type="VARCHAR(255)">
+                <constraints nullable="false"/>
+            </column>
+            <column name="WORKFLOW_EXECUTION_UUID" type="VARCHAR(100)">
```
Comment (ruchim) [inline database/migration/src/main/resources/changesets/custom_label_entry.xml]: Liquibase won't allow me to change the data type for Workflow Id. It's the FK to WorkflowMetadataSummaryEntryTable and I think it requires them to be the same data type.

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala (hunk 1 above)
Comment (ruchim) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: renamed

File: database/sql/src/main/scala/cromwell/database/slick/tables/CustomLabelEntryComponent.scala (hunk 2 above)
Comment (ruchim) [inline database/sql/src/main/scala/cromwell/database/slick/tables/CustomLabelEntryComponent.scala]: Faced some issues getting it to work so I've left it as a non Compiled query for now, but probably a good place to come back to if there are performance issues.

Comment (geoffjentry) [inline database/sql/src/main/scala/cromwell/database/slick/tables/CustomLabelEntryComponent.scala]: This is fine, but I'll note that nearly every time I see a statement like this it winds up being a performance issue and it takes a good amount of time to track it down and then someone says "oh yeah, we punted on that a few months ago"

Comment (geoffjentry) [inline database/sql/src/main/scala/cromwell/database/slick/tables/CustomLabelEntryComponent.scala]: IOW balance how much of a PITA it is now vs someone trying to debug the issue completely blind a few months down the road

File: services/src/main/scala/cromwell/services/metadata/WorkflowQueryParameters.scala (hunk 3 above)
Comment (ruchim) [inline services/src/main/scala/cromwell/services/metadata/WorkflowQueryParameters.scala]: canonicalCapitalization here just means take the query parameter keys, turn them all to lowercase and capitalize the first letter, so that these query keys can be matched up to the appropriate WorkflowQueryKey

**Comment by ruchim** [comment]
@cjllanwarne actually no, the labels hang in their own table, not in the workflow metadata summary entry table. The workflow UUID in the workflow metadata summary entry table is the FK to the table with containing labels. @geoffjentry docs added

File: services/src/main/scala/cromwell/services/metadata/WorkflowQueryKey.scala (hunk 14 above)
Comment (kshakir) [inline services/src/main/scala/cromwell/services/metadata/WorkflowQueryKey.scala]: (Check out the diff under the "Files Changed". The aforementioned magic in `.idea/inspectionProfiles/Project_Default.xml` is explicitly removed in this PR. Not sure why?)

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala [hunk 25]
```diff
@@ -112,19 +138,22 @@ trait MetadataSlickDatabase extends MetadataSqlDatabase {
...
   }
 
-  override def refreshMetadataSummaryEntries(metadataKey1: String, metadataKey2: String, metadataKey3: String,
-                                             metadataKey4: String, buildUpdatedSummary:
+  override def refreshMetadataSummaryEntries(startMetadataKey: String, endMetadataKey: String, nameMetadataKey: String,
+                                             statusMetadataKey: String, labelMetadataKey: String, buildUpdatedSummary:
                                              (Option[WorkflowMetadataSummaryEntry], Seq[MetadataEntry]) =>
                                                WorkflowMetadataSummaryEntry)
                                             (implicit ec: ExecutionContext): Future[Long] = {
+    val likeMetadataLabelKey = labelMetadataKey + "%"
```
Comment (cjllanwarne) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: ToL: is this `mysql` specific (does it work in `hsqldb`?)

File: database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowMetadataSummaryEntryComponent.scala [hunk 26]
```diff
@@ -82,10 +84,14 @@ trait WorkflowMetadataSummaryEntryComponent {
       val workflowStatusFilter = workflowStatuses.
         map(workflowStatus => workflowMetadataSummaryEntry.workflowStatus.fold(ifEmpty = exclude)(_ === workflowStatus)).
         reduceLeftOption(_ || _)
-
+      // Workflows can be queried by multiple labels, the reduceLeftOption ANDs together all workflows are associated
+      // to all the labels from the query.
+      val labelsFilter = labelKeyLabelValues.map { case (labelKey, labelValue) =>
+          existsWorkflowIdLabelKeyAndValue(workflowMetadataSummaryEntry.workflowExecutionUuid, labelKey, labelValue) }.
```
Comment (cjllanwarne) [inline database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowMetadataSummaryEntryComponent.scala]: ToL (I know you've thought about this way more than me!): This seems like a lot of work going on in Cromwell that could be compiled into a database query?

File: services/src/main/scala/cromwell/services/metadata/WorkflowQueryKey.scala (hunk 14 above)
Comment (cjllanwarne) [inline services/src/main/scala/cromwell/services/metadata/WorkflowQueryKey.scala]: Still need a space after the colon!

**Comment by cjllanwarne** [comment]
I couldn't let this go without a few final TOLs... 👍

**Comment by mcovarr** [comment]
Looks like this addresses #2106?

File: database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala (hunk 25 above)
Comment (ruchim) [inline database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala]: Confirmed to be the same in [Hsqldb](http://www.hsqldb.org/doc/guide/ch09.html)

File: engine/src/main/resources/swagger/cromwell.yaml (hunk 19 above)
Comment (ruchim) [inline engine/src/main/resources/swagger/cromwell.yaml]: Not sure how I missed this comment -- but yes. It's updated in the docs.
